dapp_runner/runner/runner.py

```diff
@@ -249,7 +259,57 @@ async def _listen_state_queue(self, service: DappService):
                 return
 
             # on a state change, we're publishing the state of the whole dapp
-            self.state_queue.put_nowait(self.dapp_state)
+            self._report_status_change()
+
+    def _report_status_change(self) -> None:
+        """Emit message with full state update to state queue."""
+
+        nodes_states = self.dapp_state
+
+        self.state_queue.put_nowait(
+            {
+                "nodes": nodes_states,
+                "app": self._get_app_state_from_nodes(
+                    len(self.dapp.nodes), self._desired_app_state, nodes_states
+                ),
+                "timestamp": utcnow_iso_str(),
+            }
+        )
+
+    # FIXME: Usage of @staticmethod is only for sake of unittests - to be removed
+    @staticmethod
+    def _get_app_state_from_nodes(
+        dapp_node_count: int,
+        desired_app_state: ServiceState,
+        nodes_states: Dict[str, Dict[int, ServiceState]],
+    ) -> ServiceState:
+        """Return general application state based on all instances states."""
+
+        # Collect nested node states into simple unique collection of state values
+        all_states = set(
+            state for node in nodes_states.values() for state in node.values()
+        )
+
+        # If we want dapp to be running handle other states as starting
+        if desired_app_state == ServiceState.running:
+            # Check node-to-state parity because of node dependency,
+            #  states gradually rolls out
+            if ({desired_app_state} == all_states) and (
+                len(nodes_states) == dapp_node_count
+            ):
+                return ServiceState.running
+
+            return ServiceState.starting
+
+        # If we want dapp to be terminated handle other states as stopping
+        if desired_app_state == ServiceState.terminated:
+            if {desired_app_state} == all_states:
+                return ServiceState.terminated
+            else:
+                return ServiceState.stopping
+
+        # In other cases return pending
+        return ServiceState.pending
```

Comment:
as mentioned above, I think there's no benefit in reporting a `pending` state in case of an app as a whole...